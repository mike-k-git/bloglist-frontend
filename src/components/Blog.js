import { useState } from "react";
import PropTypes from "prop-types";

const Blog = ({ blog, likes, currentUser, removeBlog }) => {
  const [fullInfo, setFullInfo] = useState(false);

  const toggleFullInfo = () => {
    setFullInfo(!fullInfo);
  };

  const style = {
    fontSize: 20,
    paddingTop: 10,
    paddingBottom: 10,
    paddingLeft: 20,
    border: "solid",
    marginTop: 5,
    marginBottom: 5,
    lineHeight: "1.5em",
  };

  if (fullInfo) {
    return (
      <div style={style} className="blog-item" data-cy="blog-item">
        <div>
          <span className="blog-title">{blog.title}</span>{" "}
          <span className="blog-author">{blog.author}</span>{" "}
          <button onClick={toggleFullInfo}>hide</button>
        </div>
        <div className="blog-url">{blog.url}</div>
        <div className="blog-likes">
          <span>likes {blog.likes}</span>{" "}
          <button onClick={likes} data-cy="like-submit">
            like
          </button>
        </div>
        <div className="blog-user">{blog.user.name}</div>
        {currentUser === blog.user.name && (
          <button onClick={removeBlog} data-cy="blog-remove">
            remove
          </button>
        )}
      </div>
    );
  }

  return (
    <div style={style} data-cy="blog-item">
      <span className="blog-title">{blog.title}</span>{" "}
      <span className="blog-author">{blog.author}</span>{" "}
      <button onClick={toggleFullInfo} data-cy="show-bloginfo">
        view
      </button>
    </div>
  );
};

Blog.propTypes = {
  blog: PropTypes.shape({
    likes: PropTypes.number.isRequired,
    author: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
    user: PropTypes.shape({
      username: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      id: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  likes: PropTypes.func.isRequired,
  removeBlog: PropTypes.func.isRequired,
  currentUser: PropTypes.string.isRequired,
};

export default Blog;
