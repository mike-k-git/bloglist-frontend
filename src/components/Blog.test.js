import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import userEvent from "@testing-library/user-event/";
import renderer from "react-test-renderer";
import Blog from "./Blog";

describe("<Blog />", () => {
  let container, likes, removeBlog, currentUser;

  beforeEach(() => {
    const blog = {
      title: "TDD harms architecture",
      url: "http://www.u.arizona.edu/",
      author: "Michael Chan",
      id: "1293179231",
      likes: 0,
      user: {
        username: "username",
        name: "name",
        id: "123123976123",
      },
    };
    likes = jest.fn();
    removeBlog = jest.fn();
    currentUser = "username";

    container = render(
      <Blog
        blog={blog}
        likes={likes}
        removeBlog={removeBlog}
        currentUser={currentUser}
      />
    ).container;
  });

  test("renders only blogs's title and author by default", () => {
    const titleSpan = container.querySelector(".blog-title");
    expect(titleSpan).toBeDefined();

    const authorSpan = container.querySelector(".blog-author");
    expect(authorSpan).toBeDefined();

    const urlSpan = container.querySelector(".blog-url");
    expect(urlSpan).toBeNull();

    const likesSpan = container.querySelector(".blog-likes");
    expect(likesSpan).toBeNull();
  });

  test("blog's url and number of likes are shown wthen the button clicked", async () => {
    const button = screen.getByText("view");
    await userEvent.click(button);

    const urlSpan = container.querySelector(".blog-url");
    expect(urlSpan).toBeDefined();

    const likesSpan = container.querySelector(".blog-likes");
    expect(likesSpan).toBeDefined();
  });

  test("if the like button is clicked twice, the event handler is called twice", async () => {
    const viewButton = screen.getByText("view");
    await userEvent.click(viewButton);

    const likeButton = screen.getByText("like");
    await userEvent.click(likeButton);
    await userEvent.click(likeButton);

    expect(likes.mock.calls).toHaveLength(2);
  });
});

it("renders correctly", () => {
  const tree = renderer
    .create(
      <Blog
        blog={{
          title: "TDD harms architecture",
          url: "http://www.u.arizona.edu/",
          author: "Michael Chan",
          id: "1293179231",
          likes: 0,
          user: {
            username: "username",
            name: "name",
            id: "123123976123",
          },
        }}
        likes={jest.fn()}
        removeBlog={jest.fn()}
        currentUser={"username"}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
