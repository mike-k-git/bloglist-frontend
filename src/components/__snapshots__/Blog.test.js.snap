// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  style={
    Object {
      "border": "solid",
      "fontSize": 20,
      "lineHeight": "1.5em",
      "marginBottom": 5,
      "marginTop": 5,
      "paddingBottom": 10,
      "paddingLeft": 20,
      "paddingTop": 10,
    }
  }
>
  <span
    className="blog-title"
  >
    TDD harms architecture
  </span>
   
  <span
    className="blog-author"
  >
    Michael Chan
  </span>
   
  <button
    onClick={[Function]}
  >
    view
  </button>
</div>
`;
